snap, snappy: implement the new (minmimal) kernel spec

mvo5 opened the pull request:
This branch removes the current kernel specific yaml and replaces it with the new minimal kernel.yaml spec. This spec will evolve but for now I think we can get away with this. Hardcoding vmlinuz should not break anything, its used as the filename in pi2, dragonboard, i386 and amd64 already.

pedronis commented on snap/kernel.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+// KernelYaml contains the kernel.yaml specific data
+type KernelYaml struct {
```
later if this grows complicated will want to split kernelYaml vs KernelInfo but this seems cheap for now until things are more settled

niemeyer replied:
Can we please do that change now?  Right now it's very cheap to put the foundation in place as we want it. Once it grows legs, it becomes more boring.

niemeyer replied:
Ah, sorry, it was already done. Thanks!

pedronis commented on snappy/kernel_os.go:
```diff
@@ -68,6 +68,15 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
 	}
 
+	// sanity check that we have the new kernel format
+	kmeta, err := snapf.ReadFile("meta/kernel.yaml")
```
as we discussed snapf is probably not needed here (we already mounted the snap at this point) which would simplify things, also may be that we don't Unpack in snap.File (may still be good to have around until we really know more)

pedronis replied:
from the point of view a snap.ReadKernelInfo() taking a Info or PlaceInfo and doing validation may make sense

zyga commented on snap/kernel.go:
```diff
@@ -0,0 +1,43 @@
+
+// KernelYaml contains the kernel.yaml specific data
+type KernelYaml struct {
+	Version string `yaml:"version,omitempty"`
```
Is this spec version or kernel version?

zyga commented on snap/kernel_test.go:
```diff
@@ -0,0 +1,42 @@
+var _ = Suite(&KernelYamlTestSuite{})
+
+func (s *KernelYamlTestSuite) TestValidateKernelYaml(c *C) {
+	err := snap.ValidateKernelYaml([]byte(`version: 4.4-18`))
```
Ah, disregard my earlier comment

zyga commented:
Looks good

niemeyer commented on snap/kernel.go:
```diff
@@ -0,0 +1,56 @@
+	Version string
+}
+
+func ReadKernelInfo(info *Info) (*KernelInfo, error) {
```
Some suggestions for error reporting for the function:

`const errorFormat = "cannot parse kernel snap details: %s"`

niemeyer replied:
(replaced "parse" by "read" on the message)

niemeyer commented on snap/kernel.go:
```diff
@@ -0,0 +1,56 @@
+	kernelYamlFn := filepath.Join(info.MountDir(), "meta", "kernel.yaml")
+	kmeta, err := ioutil.ReadFile(kernelYamlFn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read %q: %s", kernelYamlFn, err)
```
`return nil, fmt.Errorf(errorFormat, err)`

niemeyer commented on snap/kernel.go:
```diff
@@ -0,0 +1,56 @@
+
+	var ky kernelYaml
+	if err := yaml.Unmarshal(kmeta, &ky); err != nil {
+		return nil, fmt.Errorf("info failed to parse: %s", err)
```
`return nil, fmt.Errorf(errorFormat, err)`

niemeyer commented on snap/kernel.go:
```diff
@@ -0,0 +1,56 @@
+
+	// basic validation
+	if ky.Version == "" {
+		return nil, fmt.Errorf("missing kernel version in kernel.yaml")
```
`return nil, fmt.Errorf(errorFormat, "missing version in kernel.yaml")`

niemeyer commented on snappy/undo_test.go:
```diff
@@ -86,10 +86,11 @@ func (s *undoTestSuite) TestUndoForSetupSnapKernelUboot(c *C) {
-		{"initrd.img-4.4.0-14-generic", "initrd"},
+		{"vmlinuz", "kernel"},
+		{"initrd.img", "initrd"},
 		{"lib/modules/4.4.0-14-generic/foo.ko", "a module"},
```
Shouldn't that be "modules/..." rather than "lib/modules/..."?

niemeyer commented on snappy/undo_test.go:
```diff
@@ -86,10 +86,11 @@ func (s *undoTestSuite) TestUndoForSetupSnapKernelUboot(c *C) {
+		{"vmlinuz", "kernel"},
+		{"initrd.img", "initrd"},
 		{"lib/modules/4.4.0-14-generic/foo.ko", "a module"},
 		{"lib/firmware/bar.bin", "some firmware"},
```
Same.

niemeyer commented on snappy/kernel_os.go:
```diff
@@ -89,26 +97,24 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
-		src = filepath.Join(dstDir, src)
-		dst := filepath.Join(dstDir, dropVersionSuffix(src))
-		if err := os.Rename(src, dst); err != nil {
+	// TODO: hardcoded names until this is fully speced
```
I'd drop the comment, on the basis that this is actually our specification ATM.

niemeyer commented on snappy/kernel_os.go:
```diff
@@ -89,26 +97,24 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
-		dst := dstDir
-		if err := snapf.Unpack(src, dst); err != nil {
+
+	// TODO: hardcoded names until this is fully speced
```
Same, except the spec misses this entry. Let's talk about it.

niemeyer commented on snappy/kernel_os.go:
```diff
@@ -60,6 +55,13 @@ func removeKernelAssets(s snap.PlaceInfo, inter interacter) error {
 
+func copyAll(src, dst string) error {
+	if output, err := exec.Command("cp", "-a", src, dst).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to copy %q -> %q: %s (%q)", src, dst, err, output)
```
`"cannot copy %q to %q: %s (%s)"`

niemeyer commented on snappy/kernel_os.go:
```diff
@@ -68,6 +70,12 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
 		return fmt.Errorf("can not extract kernel assets from snap type %q", s.Type)
```
`s/can not/cannot/` here and below, please, since we're here

pedronis commented on snap/kernel.go:
```diff
@@ -0,0 +1,58 @@
+
+	// basic validation
+	if ky.Version == "" {
+		return nil, fmt.Errorf(errorFormat, err)
```
probably quick copy pasting, there's no err here,  Gustavo suggested: return nil, fmt.Errorf(errorFormat, "missing version in kernel.yaml")

mvo5 replied:
Thank you!

mvo5 commented:
@niemeyer Thanks for the detailed review. I addressed the points. The only open question is the `dtbs` directory which is not in the spec. Happy to talk about this.

niemeyer commented:
retest this please

niemeyer commented on snap/kernel.go:
```diff
@@ -0,0 +1,58 @@
+	Version string
+}
+
+const errorFormat = "cannot parse kernel snap details: %s"
```
This should be inside the function rather than being package-wide, but we can fix that later.

niemeyer commented:
LGTM

pedronis commented:
retest this please

pedronis commented on snappy/kernel_os.go:
```diff
@@ -89,26 +97,22 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
-		dst := filepath.Join(dstDir, dropVersionSuffix(src))
-		if err := os.Rename(src, dst); err != nil {
+	for _, src := range []string{
+		filepath.Join(s.MountDir(), "vmlinuz"),
```
if I understood the discussion yesterday and look at the doc it seems this one now is named kernel.img
